colossus/src/main/scala/colossus/protocols/websocket/Websocket.scala

```diff
@@ -196,11 +191,78 @@ abstract class WebsocketHandler[P <: Protocol](context: Context)
 
 }
 
-abstract class WebsocketServerHandler(serverContext: ServerContext) extends WebsocketHandler(serverContext.context) with ServerConnectionHandler {
+
+
+/**
+ * A websocket server connection handler that uses a specified sub-protocol.
+ */
+abstract class WebsocketHandler[P <: Protocol](context: Context)(implicit provider: FrameCodecProvider[P]) extends BaseWebsocketHandler(context) {
+
+  val wscodec = provider.provideCodec()
+
+  def handle: PartialFunction[P#Input, Unit]
+
+  def handleError(reason: Throwable)
+
+  def sendMessage(message: P#Output) {
+    send(wscodec.encode(message))
+  }
+
+  def processMessage(frame: Frame) {
+    frame.header.opcode match {
+      case OpCodes.Binary | OpCodes.Text => wscodec.decode(frame.payload) match {
+        case Success(obj) => handle(obj)
+        case Failure(err) => handleError(err)
+      }
+      case OpCodes.Ping => {
+        push(Frame(Header(OpCodes.Pong, false), frame.payload)){_ => {}}
+      }
+      case OpCodes.Close => {
+        disconnect()
+      }
+      case _ => {}
+    }
+  }
+}
+
+ 
+abstract class WebsocketServerHandler[P <: Protocol](serverContext: ServerContext)(implicit provider: FrameCodecProvider[P])
+extends WebsocketHandler[P](serverContext.context)(provider) with ServerConnectionHandler {
 
   implicit val namespace = serverContext.server.namespace
 
 }
 
+abstract class WebsocketInitializer(val worker: WorkerRef) {
+
+  def onConnect: ServerContext => BaseWebsocketHandler
+
+}
 
+object WebsocketServer {
+  import protocols.http._
+
+  /**
+   * Start a Websocket server on the specified port.  Since Websocket
+   * connections are upgraded from HTTP connections, this will actually start an
+   * HTTP server and react to Websocket upgrade requests on the path
+   * `upgradePath`, all other paths will 404.
+   */
+  def start(name: String, port: Int, upgradePath: String = "/")(init: WorkerRef => WebsocketInitializer)(implicit io: IOSystem) = {
+    Server.start(name, port){worker => new Initializer(worker) {
+    
+      val websockinit = init(worker)
+
+      def onConnect = ctx => new HttpService(ServiceConfig.Default,ctx) {
+        def handle = {
+          case req @ UpgradeRequest(resp) if (req.head.path == upgradePath) => {
```

Comment:
An UpgradeRequest is only a valid UpgradeRequest if the path matches the upgradePath, right?  I have mixed feelings about doing stuff like this, but in this case what if UpgradeRequest was an instance object, that has the upradePath on it, and an unapply, which can then use that reference that upgradePath.  It gets rid of this guard case, and makes UpgradeRequest unapply more complete.